keras/src/layers/layer.py

```diff
@@ -898,12 +916,21 @@ def maybe_convert(x):
                 elif self.compute_dtype != self.variable_dtype:
                     # Enter a new scope if our dtypes are "mixed".
                     new_scope = backend.AutocastScope(self.compute_dtype)
-
                 if new_scope is not None:
                     with new_scope:
-                        outputs = super().__call__(*args, **kwargs)
+                        if self._remat_mode is not None:
+                            outputs = self.rematerialized_call(
+                                super().__call__, *args, **kwargs
+                            )
+                        else:
+                            outputs = super().__call__(*args, **kwargs)
                 else:
-                    outputs = super().__call__(*args, **kwargs)
+                    if self._remat_mode is not None:
+                        outputs = self.rematerialized_call(
+                            super().__call__, *args, **kwargs
+                        )
```

Comment:
I'm confused, `rematerialized_call` is already called in `call`. Now it's also called in `__call__`, so it's called twice.

I think I prefer it in `call` so that symbolic calls are handled automatically. Meaning you'll remove it here.

Was there a reason you added it here too?

Reply:
During testing, I used mock.patch to verify that remat was being called. With wrapping just the `stateless_call`'s layer call in `rematerialization_call`-  led to errors in certain cases because remat was not invoked during the stateless_call. By wrapping the` __call__` method, I confirmed that remat was correctly called exactly once.

Reply:
Ok, so now I understand how quantization works, and I think I like the way it's done better, specifically the idea that `__call__` invokes `rematerialize_call` and not the reverse.

https://github.com/keras-team/keras/blob/master/keras/src/ops/operation.py#L38
https://github.com/keras-team/keras/blob/master/keras/src/ops/operation.py#52

Can you do the same approach? You'll have to check if the `_remat_mode` attribute exists. You'll have to add the remat case and the quantization + remat case.

Reply:
Done